core/src/main/java/org/apache/spark/api/shuffle/ShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.api.shuffle;
+
+import java.io.IOException;
+
+public interface ShuffleMapOutputWriter {
+
+  ShuffleBlockOutputStream getNextPartitionOutputStream() throws IOException;
```

Comment:
One thing that @yifeih and I discussed is the idea that getting the "next" partition output stream is easier to manage from the perspective of the caller rather than getting an output stream for a given partition ID.

The reason for this is that it's possible for the caller in theory to get a stream for a partition ID, then get another stream for a subsequent partition ID, but then subsequently write to the previously opened stream. In other words it's possible for the caller to not believe the ordering to which streams are opened and written to matters.

An API that explicitly states a notion of a "next" stream gives those who are modifying `UnsafeShuffleWriter` and similar classes a signal that they should be writing to these partition streams sequentially.